libff/algebra/curves/bls12_377/bls12_377.sage

```diff
@@ -31,6 +34,199 @@ def g1_order(curve_order):
     assert(biggest_factor[1] == 1)
     return biggest_factor[0]
 
+
+def g1_fast_subgroup_check_coefficients(g1_order, q, r):
+    """
+    Find \beta Fq with multiplicative order 3. For each element (x,y) in G1
+    with n = order or (x, y), the endomorphism \sigma(x, y) -> (\beta x, y) is
+    equivalent to [\lambda_n], where \lambda_n is a root of x^2 + x + 1 in Fn.
+
+    Find c0 and c1 s.t. c0 + c1 * \lambda_r = r
+    (i.e. [c0]P + [c1]\sigma(P) = [r]P for P in r-torsion)
+
+    Ensure that c0 + c1 * x has no roots mod any other subgroup order
+    (i.e. [c0]P + [c1]\sigma(P) does not kill members of any other torsion).
+
+    In this case, [c0]P + [c1]\sigma(P) == 0 is a fast r-torsion membership
+    check for elements of E(Fq).
+    """
+
+    print(f" [r1_0]P + [r2_1]\sigma(P) == 0")
+
+    def _find_lambda_n(Fn):
+        """
+        Return solutions of \lambda^2 + \lambda + 1 = 0 mod n
+        """
+        Fny.<y> = PolynomialRing(Fn)
+        lambda_r_poly = y^2 + y + 1
+        lambda_r_poly_roots = lambda_r_poly.roots()
+        return [root[0] for root in lambda_r_poly_roots]
+
+    def _check_lambdas_not_root_in_subgroup_size(n, c0, c1):
+        """
+        Check that no possible values of lambda_n are roots of c0 + c1*x in GF(n)
+        """
+        Fn = GF(n)
+        lambdas = _find_lambda_n(Fn)
+        for l in lambdas:
+            if 0 == Fn(c0 + c1 * l):
+                raise Exception(f"elements order {n} may be mapped to 0")
+
+    def _check_lambdas_not_root_in_subgroups(group_order, q, r, c0, c1):
+        """
+        For each prime-power factor n of group_order, perform the check that
+        [c0]P + [c1]sigma(P) does not kill the n-torsion.
+        """
+
+        def _no_element_order_4():
+            """
+            E(Fq) contains no elements of order 4 (and hence of order 2^m, m > 1).
+              E: y^2 - x^3 - 1
+              Any order 2 element has dE/dy == 0, i.e. y == 0.
+              Any order 4 element P must thereby have:
+                (P + P) = R
+                Ry == 0
+              By group laws:
+                21 * Px^6 - 5 * Px^3 + 2 == 0
```

Comment:
TODO: Clarify the origin of this.

Reply:
Sorry - this turned out to be incorrect. The correct polnomial in Px is Px^6 + 20*Px^3 - 8 which (according to Sage) does indeed yield P of order 4 (e.g 
```
[228097355113300204138531148905234651262148041026195375645000724271212049151994375092458297304264351187709081232383, 183248120370312093914300842873964025714947238860092572013797503794978351907865142483153210499531168963665210033314]
```

Thankfully, for this choice of c0, c1, even points of order 2^m do not seem to be killed by the endomorphism.
I've updated the script to reflect this https://github.com/clearmatics/libff/pull/31/commits/26fb57dc9d16f8f0f177637e8f1184ccdcb2bdf0 .  Sorry about the confusion.

Reply:
Thanks for the clarification @dtebbs